components/brave_wallet/browser/ens_resolver_task.cc

```diff
@@ -0,0 +1,495 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/ens_resolver_task.h"
+
+#include <memory>
+#include <utility>
+
+#include "base/callback_helpers.h"
+#include "base/containers/contains.h"
+#include "base/json/json_reader.h"
+#include "base/json/json_writer.h"
+#include "base/ranges/algorithm.h"
+#include "base/strings/string_util.h"
+#include "brave/components/brave_wallet/browser/brave_wallet_utils.h"
+#include "brave/components/brave_wallet/browser/eth_data_builder.h"
+#include "brave/components/brave_wallet/browser/eth_requests.h"
+#include "brave/components/brave_wallet/browser/eth_response_parser.h"
+#include "brave/components/brave_wallet/browser/json_rpc_response_parser.h"
+#include "brave/components/brave_wallet/common/eth_abi_utils.h"
+#include "brave/components/brave_wallet/common/hash_utils.h"
+#include "brave/components/brave_wallet/common/hex_utils.h"
+#include "net/base/load_flags.h"
+#include "services/network/public/cpp/resource_request.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+#include "services/network/public/cpp/simple_url_loader.h"
+#include "services/network/public/mojom/url_response_head.mojom.h"
+
+namespace brave_wallet {
+namespace {
+
+// resolve(bytes32,bytes32)
+const uint8_t kResolveBytes32Bytes32Hash[] = {0x90, 0x61, 0xb9, 0x23};
+
+absl::optional<std::vector<uint8_t>> ExtractGatewayResult(
+    const std::string& json) {
+  auto records_v =
+      base::JSONReader::Read(json, base::JSONParserOptions::JSON_PARSE_RFC);
+  if (!records_v || !records_v->is_dict()) {
+    return absl::nullopt;
+  }
+  auto* data = records_v->FindStringPath("data");
+  if (!data)
+    return absl::nullopt;
+
+  std::vector<uint8_t> result;
+  if (!PrefixedHexStringToBytes(*data, &result))
+    return absl::nullopt;
+  return result;
+}
+
+net::NetworkTrafficAnnotationTag GetNetworkTrafficAnnotationTag() {
+  return net::DefineNetworkTrafficAnnotation("ens_resolver", R"(
+      semantics {
+        sender: "ENS Resolver Task"
+        description:
+          "Fetches ENS offchain data."
+        trigger:
+          "Triggered by ENS offchain lookup."
+        data:
+          "Offchain lookup info."
+        destination: WEBSITE
+      }
+      policy {
+        cookies_allowed: NO
+        setting:
+          "You can enable or disable ENS on brave://settings/extensions page."
+        policy_exception_justification:
+          "Not implemented."
+      }
+    )");
+}
+
+std::unique_ptr<network::SimpleURLLoader> CreateLoader(
```

Comment:
This function is another reason that I would prefer we utilize api request helper if possible.

Reply:
done.